onnxruntime/core/providers/cuda/gpu_data_transfer.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/gpu_data_transfer.h"
+#include "cuda_common.h"
+
+namespace onnxruntime {
+GPUDataTransfer::GPUDataTransfer() {
+  // create streams, default is nullptr
+  streams_[kCudaStreamDefault] = nullptr;
+  CUDA_CALL_THROW(cudaStreamCreateWithFlags(&streams_[kCudaStreamCopyIn], cudaStreamNonBlocking));
+  CUDA_CALL_THROW(cudaStreamCreateWithFlags(&streams_[kCudaStreamCopyOut], cudaStreamNonBlocking));
+}
+
+bool GPUDataTransfer::CanCopy(const OrtDevice& src_device, const OrtDevice& dst_Device) const {
+  return src_device.Type() == OrtDevice::GPU 
+   || (src_device.Type() == OrtDevice::CPU && src_device.MemType() == OrtDevice::CUDA_PINNED)
+   || dst_Device.Type() == OrtDevice::GPU 
+   || (src_device.Type() == OrtDevice::CPU && src_device.MemType() == OrtDevice::CUDA_PINNED);
```

Comment:
Do you need the check on device type being CPU when MemType is CUDA_PINNED?

Reply:
Also can we have something like OrtMemType::CUDA_PINNED? It's a bit confusing to have the 'OrtDevice::' prefix used for both the device and the memory type.

---
In reply to: [296063546](https://github.com/microsoft/onnxruntime/pull/1268#discussion_r296063546) [](ancestors = 296063546)

Reply:
the OrtMemType is a good name, but unfortunately, it was somehow hacked.... I agree with you that a better name should be created. it is confusing... Will figure out.